src/ReactiveDomain.Foundation.Tests/when_using_correlated_repository.cs

```diff
@@ -0,0 +1,243 @@
+				Assert.Equal(source.MsgId, evt.CausationId);
+				Assert.Equal(source.CorrelationId, evt.CorrelationId);
+			}
+		}
+		[Fact]
+		public void updated_correlated_aggregates_inject_source_information()
+		{
+
+			ICorrelatedMessage source = MessageBuilder.New(() => new CreditAccount(_accountId, 50));
+			var retrievedAccount = _correlatedRepo.GetById<Account>(_accountId, source);
+			retrievedAccount.Credit(7);
+			retrievedAccount.Credit(13);
+			retrievedAccount.Credit(31);
+
+			var eventSource = (IEventSource)retrievedAccount;
+			var correlatedEvents = eventSource.TakeEvents().Select(evt => evt as ICorrelatedMessage).ToArray();
+			foreach (var evt in correlatedEvents) {
+				Assert.Equal(source.MsgId, evt.CausationId);
+				Assert.Equal(source.CorrelationId, evt.CorrelationId);
+			}
+		}
+
+
+		[Fact]
+		public void can_save_new_correlated_aggregates()
+		{
+			var newAccountId = Guid.NewGuid();
+			ICorrelatedMessage source = MessageBuilder.New(() => new CreateAccount(newAccountId));
+			var newAccount = new Account(newAccountId, source);
+			newAccount.Credit(7);
+			newAccount.Credit(13);
+			newAccount.Credit(31);
+			_correlatedRepo.Save(newAccount);
+
+			var retrievedAccount = _correlatedRepo.GetById<Account>(newAccountId, source);
+			Assert.NotNull(retrievedAccount);
+			Assert.Equal(51, retrievedAccount.Balance);
+			Assert.Equal(newAccountId, retrievedAccount.Id);
+		}
+
+		[Fact]
+		public void can_save_updated_correlated_aggregates()
+		{
+			ICorrelatedMessage source = MessageBuilder.New(() => new CreditAccount(_accountId, 50));
+
+			var retrievedAccount = _correlatedRepo.GetById<Account>(_accountId, source);
+			Assert.NotNull(retrievedAccount);
+			Assert.Equal(_accountId, retrievedAccount.Id);
+
+			retrievedAccount.Credit(50);
+			_correlatedRepo.Save(retrievedAccount);
+
+			var retrievedAccount2 = _correlatedRepo.GetById<Account>(_accountId, source);
+			Assert.NotNull(retrievedAccount2);
+			Assert.Equal(_accountId, retrievedAccount2.Id);
+			Assert.Equal(101, retrievedAccount.Balance);
+
+		}
+		
+
+		public class Account : AggregateRoot
+		{
+			//n.b. for infrastructure testing only not for prod or business unit tests
+			public long Balance { get; private set; }
+			//reflection constructor
+			// ReSharper disable once UnusedMember.Local
+			private Account()
+			{
+				Register<AccountCreated>(evt => Id = evt.AccountId);
+				Register<AccountCredited>(evt => { Balance += evt.Amount; });
+			}
+			public Account(Guid id, ICorrelatedMessage source) : this()
+			{
+				((ICorrelatedEventSource)this).Source = source;
+				Ensure.NotEmptyGuid(id, "id");
+				Raise(new AccountCreated(id));
+			}
+
+			public void Credit(uint amount)
+			{
+				Raise(new AccountCredited(Id, amount));
+			}
+		}
+		public class CreateAccount : Command
+		{
+			public readonly Guid AccountId;
+			public CreateAccount(Guid accountId)
+			{
+				AccountId = accountId;
+			}
+		}
+		public class AccountCreated : ICorrelatedMessage
```

Comment:
Can this inherit from `Event` instead to avoid manually implementing `ICorrelatedMessage`?

Reply:
Changed to use Event here

Reply:
the cache and the repository are designed to be composable and are tested separately